wagtail/wagtailcore/models.py

```diff
@@ -40,8 +39,86 @@
 
 PAGE_TEMPLATE_VAR = 'page'
 
+SITE_CACHE = {}
+REQUEST_SITE_MATCH_CACHE = {}
+
 
 class SiteManager(models.Manager):
+
+    @staticmethod
+    def get_hostname_and_port_from_request(request):
+        """Reliably extract 'hostname' and 'port' values from the supplied
+        ``HttpRequest`` instance (without raising exceptions), and return them
+        as a tuple.
+        """
+        try:
+            hostname = request.get_host().split(':')[0]
+        except KeyError:
+            hostname = None
+        try:
+            port = request.get_port()
+        except KeyError:
+            port = None
+        return hostname, port
+
+    @staticmethod
+    def clear_site_cache():
+        global SITE_CACHE
+        global REQUEST_SITE_MATCH_CACHE
+        SITE_CACHE = {}
+        REQUEST_SITE_MATCH_CACHE = {}
+
+    def _get_by_id(self, site_id):
+        if site_id not in SITE_CACHE:
+            site = self.get(pk=site_id)
+            SITE_CACHE[site_id] = site
+        return SITE_CACHE[site_id]
+
+    def _cache_and_return_match(self, site, match_cache_key):
+        SITE_CACHE[site.pk] = site
+        REQUEST_SITE_MATCH_CACHE[match_cache_key] = site.pk
+        return SITE_CACHE[site.pk]
+
+    def get_default(self):
+        """Returns the 'default' ``Site`` or raise an exception in no site is
+        set as the default. The result is cached."""
+        key = 'default'
+        if key in REQUEST_SITE_MATCH_CACHE:
+            return self._get_by_id(REQUEST_SITE_MATCH_CACHE['default'])
+
+        match = self.get(is_default_site=True)
```

Comment:
What if multiple sites have `is_default_site=True`?
I suggest to change to `self.filter(is_default_site=True).first()`.

Reply:
This one I'm a little unsure about. I know site data can be altered in ways that bypasses this, but it seems that the model assumes there should only be one default site:
https://github.com/wagtail/wagtail/blob/master/wagtail/wagtailcore/models.py#L145

I fully understand that doing as you propose would cater for every scenario, but does that mean that the code contradicts itself slightly?

Reply:
@BertrandBordage See also: http://docs.wagtail.io/en/v1.13.1/reference/pages/model_reference.html#wagtail.wagtailcore.models.Site.is_default_site

Reply:
Yeah, a bit tricky to decide. It’s either being tolerant and strict.
We can keep it strict as you wrote, but maybe it would be better to explain what’s happening, like catching the error and reraising it with a more explicit message:
```
MultipleObjectsReturned: Multiple default sites found, only one allowed.
```